File: src/calendar.ts
import type { FastMCP } from 'fastmcp';

import { z } from 'zod';

import type { WahooClient } from '../client.js';

/**
 * Register calendar management tools
 */
export const registerCalendarTools = (
  mcp: FastMCP,
  client: WahooClient,
  ensureAuth: () => void
) => {
  // Tool: get_calendar
  mcp.addTool({
    annotations: {
      readOnlyHint: true,
      title: 'Get Calendar Workouts',
    },
    description:
      'Get planned workouts from Wahoo SYSTM calendar for a date range. Returns workout name, type, duration, planned date, and basic details.',
    execute: async (args) => {
      ensureAuth();
      const workouts = await client.getCalendarWorkouts(args.start_date, args.end_date);
      return JSON.stringify(workouts, null, 2);
    },
    name: 'get_calendar',
    parameters: z.object({
      end_date: z.string().describe('End date in YYYY-MM-DD format'),
      start_date: z.string().describe('Start date in YYYY-MM-DD format'),
    }),
  });

  // Tool: schedule_workout
  mcp.addTool({
    annotations: {
      destructiveHint: true,
      readOnlyHint: false,
      title: 'Schedule Workout',
    },
    description: 'Schedule a workout from the library to your calendar for a specific date.',
    execute: async (args) => {
      ensureAuth();
      const agendaId = await client.scheduleWorkout(
        args.content_id,
        args.date,
        args.time_zone || 'UTC'
      );
      return JSON.stringify(
        {
          agendaId,
          date: args.date,
          success: true,
          timezone: args.time_zone || 'UTC',
        },
        null,
        2
      );
    },
    name: 'schedule_workout',
    parameters: z.object({
      content_id: z
        .string()
        .describe(
          'Workout content ID from library search results (use the id field, not workoutId)'
        ),
      date: z.string().describe('Date to schedule the workout (YYYY-MM-DD format)'),
      time_zone: z
        .string()
        .optional()
        .describe(
          'Timezone for the workout (default: UTC). Example: Europe/Lisbon, America/New_York'
        ),
    }),
  });

  // Tool: reschedule_workout
  mcp.addTool({
    annotations: {
      destructiveHint: true,
      idempotentHint: true,
      readOnlyHint: false,
      title: 'Reschedule Workout',
    },
    description: 'Move a scheduled workout to a different date.',
    execute: async (args) => {
      ensureAuth();
      await client.rescheduleWorkout(args.agenda_id, args.new_date, args.time_zone || 'UTC');
      return JSON.stringify(
        {
          agendaId: args.agenda_id,
          message: 'Workout rescheduled successfully',
          newDate: args.new_date,
          success: true,
          timezone: args.time_zone || 'UTC',
        },
        null,
        2
      );
    },
    name: 'reschedule_workout',
    parameters: z.object({
      agenda_id: z.string().describe('Agenda ID from get_calendar or schedule_workout'),
      new_date: z.string().describe('New date for the workout (YYYY-MM-DD format)'),
      time_zone: z
        .string()
        .optional()
        .describe('Timezone for the rescheduled workout (default: UTC)'),
    }),
  });

  // Tool: remove_workout
  mcp.addTool({
    annotations: {
      destructiveHint: true,
      idempotentHint: true,
      readOnlyHint: false,
      title: 'Remove Workout',
    },
    description: 'Remove a scheduled workout from your calendar.',
    execute: async (args) => {
      ensureAuth();
      await client.removeWorkout(args.agenda_id);
      return JSON.stringify(
        {
          agendaId: args.agenda_id,
          message: 'Workout removed successfully',
          success: true,
        },
        null,
        2
      );
    },
    name: 'remove_workout',
    parameters: z.object({
      agenda_id: z.string().describe('Agenda ID from get_calendar or schedule_workout'),
    }),
  });
};

File: src/wahoo_systm_mcp/tools/calendar.py
"""Calendar tools for Wahoo SYSTM MCP."""

from fastmcp import Context, FastMCP

from wahoo_systm_mcp.models import (
    RemoveWorkoutResultOut,
    RescheduleWorkoutResultOut,
    ScheduleWorkoutResultOut,
    UserPlanItemOut,
)
from wahoo_systm_mcp.server.lifecycle import get_client


async def get_calendar(
    ctx: Context, start_date: str, end_date: str, time_zone: str = "UTC"
) -> list[UserPlanItemOut]:
    """Get planned workouts from Wahoo SYSTM calendar for a date range.

    Returns workout name, type, duration, planned date, and basic details.

    Args:
        start_date: Start date in YYYY-MM-DD format
        end_date: End date in YYYY-MM-DD format
        time_zone: Timezone for the calendar (default: UTC)

    """
    client = get_client(ctx)
    workouts = await client.get_calendar(start_date, end_date, time_zone)
    return [UserPlanItemOut.model_validate(w.model_dump()) for w in workouts]


async def schedule_workout(
    ctx: Context,
    content_id: str,
    date: str,
    time_zone: str = "UTC",
) -> ScheduleWorkoutResultOut:
    """Schedule a workout from the library to your calendar for a specific date.

    Args:
        content_id: Workout content ID from library search results (use the id field, not workoutId)
        date: Date to schedule the workout (YYYY-MM-DD format)
        time_zone: Timezone for the workout (default: UTC). Example: Europe/Lisbon, America/New_York

    """
    client = get_client(ctx)
    agenda_id = await client.schedule_workout(content_id, date, time_zone)
    return ScheduleWorkoutResultOut(
        success=True,
        agenda_id=agenda_id,
        date=date,
        time_zone=time_zone,
    )


async def reschedule_workout(
    ctx: Context,
    agenda_id: str,
    new_date: str,
    time_zone: str = "UTC",
) -> RescheduleWorkoutResultOut:
    """Move a scheduled workout to a different date.

    Args:
        agenda_id: Agenda ID from get_calendar or schedule_workout
        new_date: New date for the workout (YYYY-MM-DD format)
        time_zone: Timezone for the rescheduled workout (default: UTC)

    """
    client = get_client(ctx)
    await client.reschedule_workout(agenda_id, new_date, time_zone)
    return RescheduleWorkoutResultOut(
        success=True,
        message="Workout rescheduled successfully",
        agenda_id=agenda_id,
        new_date=new_date,
        time_zone=time_zone,
    )


async def remove_workout(ctx: Context, agenda_id: str) -> RemoveWorkoutResultOut:
    """Remove a scheduled workout from your calendar.

    Args:
        agenda_id: Agenda ID from get_calendar or schedule_workout

    """
    client = get_client(ctx)
    await client.remove_workout(agenda_id)
    return RemoveWorkoutResultOut(
        success=True,
        message="Workout removed successfully",
        agenda_id=agenda_id,
    )


def register(app: FastMCP) -> None:
    """Register calendar tools with the FastMCP app."""
    app.tool(get_calendar)
    app.tool(schedule_workout)
    app.tool(reschedule_workout)
    app.tool(remove_workout)
